### ISSUE TITLE
Should a @param-out be considered a contract for sub-classes?

### ISSUE THREAD

**Issue by jsiefer** [vimeo/psalm#4885]
See https://psalm.dev/r/214d87922e

Should a `@param-out` be considered a contract for sub-classes?

Best explained in the example:
```php
<?php
abstract class A
{
    /**
     * @param-out stdClass $bar
     */
    abstract public function foo(stdClass &$bar = null): int;          //  Variable $bar is limited to values of type stdClass because it is passed by reference, null|stdClass type found
}

class B extends A
{
    /**
     * @param-out stdClass $bar
     */
    public function foo(stdClass &$bar = null): int
    {
        $bar = $bar ?? new stdClass();

        return 0;
    }
}

class C extends A
{
    /**
     * @param-out stdClass|null $bar                                      // Should complain here?
     */
    public function foo(stdClass &$bar = null): int 
    {
        $bar = null;

        return 0;
    }
}
```

At the moment psalm reports that `A::foo()` `$bar` may be null. Since it can not check the implementation, I suppose that is fair. 

But I wonder if it should become a contract, and then complain at `C::foo` for violating it.

**Comment by orklah** [comment]
I'm not actually sure what LSP would tell about that. It's a parameter but it acts more like a return type. Maybe this exists in other languages that solved this in their static analysis tool?
